### PR TITLE
media-gfx/hydrus: Update to 442 + upnpc fix

### PR DESCRIPTION
**Pull request by vaartis**
Signed-off-by: Ekaterina Vaartis <vaartis@kotobank.ch>

### CODE REVIEW

**Comment by vaartis** [comment]
Updated the PR to also include a fix to a problem i discovered with UPnPC, because it seems that gentoo has a name for it that the package didn't expect. Fixed it in a way that should work in other distros too, patch submitted upstream.

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+		cloudscraper? ( dev-python/cloudscraper[${PYTHON_MULTI_USEDEP}] )
+		ffmpeg? ( media-video/ffmpeg )
+		lz4? ( dev-python/lz4[${PYTHON_MULTI_USEDEP}] )
+		mpv? (
+			 media-video/mpv[libmpv,${PYTHON_MULTI_USEDEP}]
+			 dev-python/python-mpv[${PYTHON_MULTI_USEDEP}]
+		)
+		socks? (
+				|| ( dev-python/requests[socks5,${PYTHON_MULTI_USEDEP}]
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:55]: Too much indent.

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+src_compile() {
+	python_optimize "${S}"
+}
+
+src_test() {
+	# The tests use unittest, but are run with a custom runner script.
+	# QT_QPA_PLATFORM is required to make them run without X
+	export QT_QPA_PLATFORM=offscreen
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:111]:
```suggestion
	local -x QT_QPA_PLATFORM=offscreen
```

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	rm -r hydrus/test/ || die
+	rm -r static/testing/ || die
+}
+
+src_prepare() {
+	default
+
+	# If tests will run, leave the files until tests are run.
+	# They will be deleted before installing the package.
+	if ! use test; then
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:90]: I don't really think there's a point in making it conditional. Just remove them in install phase unconditionally.

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+src_test() {
+	# The tests use unittest, but are run with a custom runner script.
+	# QT_QPA_PLATFORM is required to make them run without X
+	export QT_QPA_PLATFORM=offscreen
+	"${EPYTHON}" "${S}/test.py" || die "Tests failed"
+}
+
+src_install() {
+	local DOC="${EPREFIX}/usr/share/doc/${PF}"
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:116]: Uppercase is for global vars.

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	elog "Hydrus includes an excellent manual, that can either be viewed at"
+	elog "${DOC}/html/help/index.html"
+	elog "or accessed through the hydrus help menu."
+
+	mv "help my client will not boot.txt" "help_my_client_will_not_boot.txt" || die
+
+	DOCS=(COPYING README.md Readme.txt help_my_client_will_not_boot.txt db/)
+	HTML_DOCS=("${S}"/help/)
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:124]: `local` both.

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+	if use test; then
+		# Delete files only needed for tests now. No need to install them.
+		# If the tests didn't run, the files have been deleted already.
+		delete_test_files
+	fi
+
+	# These files are copied into DOC
+	rm COPYING README.md Readme.txt help_my_client_will_not_boot.txt || die
+	rm -r help/ db/ || die
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:135]:
```suggestion
	rm -r "${DOCS[@]}" "${HTML_DOCS[@]}" || die
```

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	# The program expects to find documentation here, so add a symlink to DOC
+	ln -s "${DOC}/html/help" help || die
+
+	insopts -m0755
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:139]: This looks somewhat wrong. Are all fixes really executable?

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	insinto /opt/hydrus
+	doins -r "${S}"/*
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:141]:
```suggestion
	doins -r "${S}"/.
```

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+	exeinto /usr/bin
+
+	sed "s/python/${EPYTHON}/" "${FILESDIR}/hydrus-server" > "${T}/hydrus-server" || die
+	sed "s/python/${EPYTHON}/" "${FILESDIR}/hydrus-client" > "${T}/hydrus-client" || die
+
+	python_doexe "${T}/hydrus-server"
+	python_doexe "${T}/hydrus-client"
+
+	make_desktop_entry "hydrus-client" "Hydrus Client" "/opt/hydrus/static/hydrus_non-transparent.png"\
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:151]:
```suggestion
	make_desktop_entry "hydrus-client" "Hydrus Client" "/opt/hydrus/static/hydrus_non-transparent.png" \
```

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	insopts -m0755
+	insinto /opt/hydrus
+	doins -r "${S}"/*
+
+	exeinto /usr/bin
+
+	sed "s/python/${EPYTHON}/" "${FILESDIR}/hydrus-server" > "${T}/hydrus-server" || die
+	sed "s/python/${EPYTHON}/" "${FILESDIR}/hydrus-client" > "${T}/hydrus-client" || die
+
+	python_doexe "${T}/hydrus-server"
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:148]:
```suggestion
	python_newexe - hydrus-server < <(sed ...)
```

**Review by mgorny — CHANGES_REQUESTED** [review]

File: media-gfx/hydrus/hydrus-442.ebuild [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{7..9} )
+PYTHON_REQ_USE="sqlite"
+
+inherit python-single-r1 desktop optfeature
+
+DESCRIPTION="A booru-like media organizer for the desktop"
+HOMEPAGE="http://hydrusnetwork.github.io/hydrus/ https://github.com/hydrusnetwork/hydrus"
```
Comment (mgorny) [inline media-gfx/hydrus/hydrus-442.ebuild:12]: https

**Review by mgorny — CHANGES_REQUESTED** [review]
Unfortunately, it doesn't seem to work for me:

```
>>> Test phase: media-gfx/hydrus-442
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/qtpy/__init__.py", line 204, in <module>
    raise ImportError # __version__ as PYSIDE_VERSION  # analysis:ignore
ImportError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/portage/media-gfx/hydrus-442/work/hydrus-442/test.py", line 7, in <module>
    from hydrus import hydrus_test
  File "/tmp/portage/media-gfx/hydrus-442/work/hydrus-442/hydrus/hydrus_test.py", line 3, in <module>
    from hydrus.client.gui import QtPorting as QP
  File "/tmp/portage/media-gfx/hydrus-442/work/hydrus-442/hydrus/client/gui/QtPorting.py", line 22, in <module>
    import qtpy
  File "/usr/lib/python3.8/site-packages/shiboken2/files.dir/shibokensupport/__feature__.py", line 142, in _import
    return original_import(name, *args, **kwargs)
  File "/usr/lib/python3.8/site-packages/qtpy/__init__.py", line 210, in <module>
    raise PythonQtError('No Qt bindings could be found')
qtpy.PythonQtError: No Qt bindings could be found
```

```
[I] dev-python/QtPy
     Available versions:  1.9.0-r3^t (~)1.9.0-r5^t{xpak:3} {declarative designer gui help location multimedia network opengl positioning printsupport +pyqt5 pyside2 sensors serialport speech sql svg test testlib webchannel webengine websockets widgets x11extras xml xmlpatterns PYTHON_TARGETS="python3_10 python3_8 python3_9"}
     Installed versions:  1.9.0-r5^t{xpak:3}(11:08:40 04.06.2021)(gui opengl pyqt5 pyside2 svg xml -declarative -designer -help -location -multimedia -network -positioning -printsupport -sensors -serialport -speech -sql -test -testlib -webchannel -webengine -websockets -widgets -x11extras -xmlpatterns PYTHON_TARGETS="python3_8 -python3_10 -python3_9")
     Homepage:            https://github.com/spyder-ide/qtpy/
     Description:         Abstraction layer on top of PyQt5 and PySide2 and additional custom QWidgets
```

File: media-gfx/hydrus/hydrus-442.ebuild (hunk 11 above)
Comment (vaartis) [reply media-gfx/hydrus/hydrus-442.ebuild]: Fixed this one, for some reason it seems that QtPy/pyside2 didn't pull the python 3.9 version for you? I'm pretty sure I put the usedeps on both, that's strange.

**Comment by mgorny** [comment]
I guess my pyside2 is broken:

    ImportError: /usr/lib64/libpyside2-python3.8.so.5.15: undefined symbol: _ZN8Shiboken6PyName9select_idEv

Unfortunately, I can't rebuild it because of https://bugs.gentoo.org/749330

**Comment by vaartis** [comment]
> I guess my pyside2 is broken:
> 
> ```
> ImportError: /usr/lib64/libpyside2-python3.8.so.5.15: undefined symbol: _ZN8Shiboken6PyName9select_idEv
> ```
> 
> Unfortunately, I can't rebuild it because of https://bugs.gentoo.org/749330

Ah.. guess it builds for me because I have clang installed too.

**Comment by vaartis** [comment]
> I guess my pyside2 is broken:
> 
> ```
> ImportError: /usr/lib64/libpyside2-python3.8.so.5.15: undefined symbol: _ZN8Shiboken6PyName9select_idEv
> ```
> 
> Unfortunately, I can't rebuild it because of https://bugs.gentoo.org/749330

Ah.. guess it builds for me because I have clang installed too.

**Comment by juippis** [comment]
```
Ran 4 tests in 13.218s

FAILED (failures=1)
This was version 442
>>> Completed testing media-gfx/hydrus-442
```
https://dev.gentoo.org/~juippis/tmp/hydruses-pr-21100.tar.xz
(Note I disabled python-3.10 globally at some point because emerge couldn't solve the REQUIRED_USE mess)

**Comment by vaartis** [comment]
> ```
> Ran 4 tests in 13.218s
> 
> FAILED (failures=1)
> This was version 442
> >>> Completed testing media-gfx/hydrus-442
> ```
> 
> https://dev.gentoo.org/~juippis/tmp/hydruses-pr-21100.tar.xz
> (Note I disabled python-3.10 globally at some point because emerge couldn't solve the REQUIRED_USE mess)

Solved the errors from the logs by explicitly requiring png & jpeg on opencv.

**Comment by vaartis** [comment]
Also patched the tests so they exit correctly when there's an error

**Comment by vaartis** [comment]
ping

File: media-gfx/hydrus/hydrus-441-r1.ebuild [hunk 12]
```diff
@@ -70,26 +70,14 @@ BDEPEND="
 "
 
 PATCHES=(
+	"${FILESDIR}/upnpc.patch"
 	"${FILESDIR}/userpath-in-local-share.patch"
+	"${FILESDIR}/test-exitcode.patch"
```
Comment (juippis) [inline media-gfx/hydrus/hydrus-441-r1.ebuild:75]: You forgot to push this patch file :( 
```
 * Applying test-exitcode.patch ...
/var/tmp/portage/media-gfx/hydrus-441-r1/temp/environment: line 1013: /var/tmp/portage/media-gfx/hydrus-441-r1/files/test-exitcode.patch: No such file or directory
/var/tmp/portage/media-gfx/hydrus-441-r1/temp/environment: line 1016: /var/tmp/portage/media-gfx/hydrus-441-r1/files/test-exitcode.patch: No such file or directory
```

Comment (vaartis) [reply media-gfx/hydrus/hydrus-441-r1.ebuild:75]: You're right, I did. Pushed it just now

**Comment by mgorny** [comment]
Ok, I've managed to rebuild pyside2, so I can merge this now.

**Pull request closed without merging**